### PR TITLE
[TH2-4299] Manual Acknowledgement refinement in common library

### PR DESCRIPTION
**Pull request by Xanclry**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java [hunk 1]
```diff
@@ -128,8 +129,7 @@ public void start() throws Exception {
                                         try {
                                             value = valueFromBytes(delivery.getBody());
                                         } catch (Exception e) {
-                                            LOGGER.error("Couldn't parse delivery. Confirm message received", e);
-                                            confirmation.confirm();
+                                            confirmProcessed.reject();
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java:132]:
```suggestion
                                            LOGGER.error("Couldn't parse delivery. Reject message received", e);
                                            confirmProcessed.reject();
```

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java [hunk 2]
```diff
@@ -239,7 +240,7 @@ protected void handle(String consumeTag, Delivery delivery, T value, Confirmatio
         } catch (Exception e) {
             LOGGER.error("Can not parse value from delivery for: {}", consumeTag, e);
             try {
-                confirmation.confirm();
+                confirmation.reject();
             } catch (IOException ex) {
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java:244]: Maybe remove this try catch as above reject

File: src/main/kotlin/com/exactpro/th2/common/schema/message/DeliveryMetadata.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2020-2022 Exactpro (Exactpro Systems Limited)
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/common/schema/message/DeliveryMetadata.kt:2]:
```suggestion
 * Copyright 2022 Exactpro (Exactpro Systems Limited)
```

File: src/main/kotlin/com/exactpro/th2/common/schema/message/ManualAckDeliveryCallback.kt [hunk 4]
```diff
@@ -20,17 +20,14 @@ import com.rabbitmq.client.Delivery
...
-    /**
-     * Called when a delivery from queue is received
-     * @param consumerTag the _consumer_ tag associated with the consumer
-     * @param delivery the delivered message
-     * @param confirmProcessed the action that should be invoked when the message can be considered as processed
-     */
+
     @Throws(IOException::class)
-    fun handle(consumerTag: String, delivery: Delivery, confirmProcessed: Confirmation)
+    fun handle(deliveryMetadata: DeliveryMetadata, delivery: Delivery, confirmProcessed: Confirmation)
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/common/schema/message/ManualAckDeliveryCallback.kt:25]: Please revert javadoc

File: src/main/kotlin/com/exactpro/th2/common/schema/message/impl/OnlyOnceConfirmation.kt [hunk 5]
```diff
@@ -34,6 +34,14 @@ class OnlyOnceConfirmation private constructor(
         }
     }
 
+    override fun reject() {
+        if (called.compareAndSet(false, true)) {
+            delegate.reject()
+        } else {
+            LOGGER.warn { "Confirmation '$id' invoked more that one time" }
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/common/schema/message/impl/OnlyOnceConfirmation.kt:41]:
```suggestion
            LOGGER.warn { "Confirmation or rejection '$id' invoked more that one time" }
```

Comment (Nikita-Smirnov-Exactpro) [reply src/main/kotlin/com/exactpro/th2/common/schema/message/impl/OnlyOnceConfirmation.kt:41]: The seam change is required for the confirm method

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java (hunk 2 above)
Comment (Xanclry) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java:244]: try-catch can't be removed, reject() throws IOException. The reject() above wrapped into another try-catch

Comment (Nikita-Smirnov-Exactpro) [reply src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/AbstractRabbitSubscriber.java:245]: the `protected void handle(String consumeTag, Delivery delivery, T value, Confirmation confirmation)` method can throw IOException

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**